ERCS/erc-7579.md

```diff
@@ -0,0 +1,351 @@
+
+### Modules
+
+This standard separates modules into the following different types that each has a unique and incremental identifier, which SHOULD be used by accounts, modules and other entities to identify the module type:
+
+- Validation (type id: 1)
+- Execution (type id: 2)
+- Fallback (type id: 3)
+- Hooks (type id: 4)
+
+Note: A single module can be of multiple types.
+
+Modules MUST implement the following interface, which is used by smart accounts to install and uninstall modules:
+
+```solidity
+interface IModule {
+    function onInstall(bytes calldata data) external;
+    function onUninstall(bytes calldata data) external;
+    function isModuleType(uint256 typeID) external view returns(bool);
+}
+```
+
+Modules MUST revert if `onInstall` or `onUninstall` was unsuccessful.
+
+#### Validators
+
+Validators MUST implement the `IModule` and the `IValidator` interface and have module type id: `1`.
+
+```solidity
+interface IValidator {
+    /**
+    * MUST validate that the signature is a valid signature of the userOpHash, and SHOULD return ERC-4337's SIG_VALIDATION_FAILED (and not revert) on signature mismatch
+    */
+    function validateUserOp(UserOperation calldata userOp, bytes32 userOpHash) external returns (uint256);
+
+    /**
+    * @dev `sender` is the address that sent the ERC-1271 request to the smart account
+    *
+    * MUST return the ERC-1271 `MAGIC_VALUE` if the signature is valid
+    * MUST NOT modify state
+    */
+    function isValidSignatureWithSender(address sender, bytes32 hash, bytes calldata signature) external view returns (bytes4);
+}
+```
+
+#### Executors
+
+Executors MUST implement the `IModule` interface and have module type id: `2`.
+
+#### Fallback Handlers
+
+Fallback handlers MUST implement the `IModule` interface and have module type id: `3`.
+
+Fallback handlers that implement authorization control, MUST NOT rely on `msg.sender` for authorization control but MUST use ERC-2771 `_msgSender()` instead.
+
+#### Hooks
+
+Hooks MUST implement the `IModule` and the `IHook` interface and have module type id: `4`.
+
+```solidity
+interface IHook {
+    /**
+    *  MAY return arbitrary data in the `hookData` return value
+    */
+    function preCheck(address msgSender, bytes calldata msgData) external returns (bytes memory hookData);
+
+    /**
+    * MAY validate the `hookData` to validate transaction context of the `preCheck` function
+    */
+    function postCheck(bytes calldata hookData) external returns (bool success);
+}
+```
+
+## Rationale
+
+### Standardization
+
+As mentioned above, there are several reasons for why standardizing smart accounts is very beneficial to the ecosystem. The most important of these are:
+
+- Interoperability for modules to be used across different smart accounts
+- Interoperability for smart accounts to be used across different wallet applications and sdks
+- Preventing significant vendor lock-in for smart account users
+
```

Comment:
The content in your rationale section should explain decisions made within the proposal itself, not the proposal as a whole. That content belongs in your motivation section.